dstrf/_dstrf.py

```diff
@@ -82,57 +84,96 @@ def dstrf(meg, stim, lead_field, noise, tstart=0, tstop=0.5, nlevels=1,
+    ``y2`` are responses to stimuli ``x1`` and ``x2``, respoectively::
+
+        dstrf([y1, y2], [x1, x2], fwd, cov)
+
+    And with multiple predictors::
+
+        dstrf([y1, y2], [[x1_attended, x1_unattended], [x2_attended, x2_unattended]], fwd, cov)
+
     """
-    # noise covariance
-    if isinstance(noise, NDVar):
-        er = noise.get_data(('sensor', 'time'))
-        noise_cov = np.dot(er, er.T) / er.shape[1]
-    elif isinstance(noise, np.ndarray):
-        if (noise.ndim == 2) and (noise.shape[0] == noise.shape[0]):
-            noise_cov = noise
-        else:
-            raise ValueError(f'For noise as ndarray, noise dim1={noise.shape[0]} should'
-                             f'match dim2={noise.shape[0]}')
+    # normalize=True defaults to 'l2'
+    if normalize is True:
+        normalize = 'l2'
+    elif isinstance(normalize, str):
+        if normalize not in ('l1', 'l2'):
+            raise ValueError(f"normalize={normalize!r}, need bool or 'l1' or 'l2'")
     else:
-        raise NotImplementedError
-
-    # Initialize `REG_Data` instance with desired properties
-    ds = REG_Data(tstart, tstop, nlevels)
+        raise TypeError(f"normalize={normalize!r}, need bool or str")
 
     # data copy?
-    if in_place is None:
-        in_place = False
     if not isinstance(in_place, bool):
         raise TypeError(f"in_place={in_place!r}, need bool or None")
 
     # Call `REG_Data.add_data` once for each contiguous segment of MEG data
-    for r, s in iter_data(meg, stim):
-        if not in_place:
-            s = s.copy()
-            r = r.copy()
-
-        if normalize:  # screens False, None
-            s -= s.mean('time')
-            if isinstance(normalize, bool):  # normalize=True defaults to 'l2'
-                normalize = 'l2'
-            if isinstance(normalize, str):
+    ds = REG_Data(tstart, tstop, nlevels)
+    for r, ss in iter_data(meg, stim):
+        if normalize:
+            if not in_place:
+                ss = [s.copy() for s in ss]
+            for s in ss:
+                s -= s.mean('time')
                 if normalize == 'l2':
-                    s_scale = (s.x ** 2).mean(-1) ** 0.5
+                    s_scale = (s ** 2).mean('time') ** 0.5
                 elif normalize == 'l1':
-                    s_scale = np.abs(s.x).mean(-1)
+                    s_scale = s.abs().mean('time')
                 else:
-                    raise ValueError(f"normalize={normalize!r}, need bool or \'l1\' or \'l2\'")
-            else:
-                raise TypeError(f"normalize={normalize!r}, need bool or str")
+                    raise RuntimeError(f"normalize={normalize!r}")
+                s /= s_scale
+        ds.add_data(r, ss)
+    # TODO: make this less hacky when fixing normalization (iter_data() always turns stim into lists)
+    ds._stim_is_single = isinstance(stim, NDVar) if isinstance(meg, NDVar) else isinstance(stim[0], NDVar)

```

Comment:
Shouldn't we do the normalization in `REG_Data` after all data is added, make sure the same normalization factor is used for all trials, and store the normalization factor so that we can then rescale `h` to match the scale of the original stimuli?